dask_match/core.py

```diff
@@ -616,7 +625,245 @@ def optimize_expr(expr):
+        seen = set()
+        stack = [expr]
+        dependents = defaultdict(set)
+        dependencies = {}
+        while stack:
+            next = stack.pop()
+
+            if next._name in seen:
+                continue
+            seen.add(next._name)
+
+            if isinstance(next, Fusable):
+                dependencies[next] = set()
+                if next not in dependents:
+                    dependents[next] = set()
+
+            for operand in next.operands:
+                if isinstance(operand, Expr):
+                    stack.append(operand)
+                    if isinstance(operand, Fusable):
+                        if next in dependencies:
+                            dependencies[next].add(operand)
+                        dependents[operand].add(next)
+
+        # Traverse each "root" until we find a fusable sub-group.
+        # Here we use root to refer to a Fusable Expr node that
+        # has no Fusable dependents
+        roots = [
+            k
+            for k, v in dependents.items()
+            if v == set() or all(not isinstance(_expr, Fusable) for _expr in v)
+        ]
+        while roots:
+            root = roots.pop()
+            seen = set()
+            stack = [root]
+            group = []
+            while stack:
+                next = stack.pop()
+
+                if next._name in seen:
+                    continue
+                seen.add(next._name)
+
+                group.append(next)
+                for dep in dependencies[next]:
+                    if not (dependents[dep] - set(stack) - set(group)):
+                        # All of deps dependents are contained
+                        # in the local group (or the local stack
+                        # of expr nodes that we know we will be
+                        # adding to the local group)
+                        stack.append(dep)
+                    elif dep not in roots and dependencies[dep]:
+                        # Couldn't fuse dep, but we may be able to
+                        # use it as a new root on the next pass
+                        roots.append(dep)
+
+            # Replace fusable sub-group
+            if len(group) > 1:
+                group_deps = []
+                local_names = [_expr._name for _expr in group]
+                for _expr in group:
+                    group_deps += [
+                        operand
+                        for operand in _expr._subgraph_dependencies()
+                        if operand._name not in local_names
+                    ]
+                to_replace = {group[0]: FusedExpr(group, *group_deps)}
+                return replace_nodes(expr, to_replace), not roots
+
+        # Return original expr if no fusable sub-groups were found
+        return expr, True
+
+    while True:
+        original_name = expr._name
+        expr, done = _fusion_pass(expr)
+        if done or expr._name == original_name:
+            break
+
+    return expr
```

Comment:
I'm not reviewing this logic, but I think that it would be good to think of and write down some more complex test cases.